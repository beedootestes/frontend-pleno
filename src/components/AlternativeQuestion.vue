<template>
  <div
    @click="emit('click', $event)"
    :class="[{selected: selected}, isCorrect]"
    class="question-item">
    <div>{{ alternative.text }}</div>

    <div>
      <CheckCircleIcon class="correct-icon" v-if="isCorrect === 'correct'" :size="45" fill-color="#228B21" />
      <AlphaXCircleIcon v-else-if="isCorrect === 'incorrect'" :size="45" fill-color="#FF0000" />
    </div>
  </div>
</template>

<script setup lang="ts">
  import AlphaXCircleIcon from 'vue-material-design-icons/AlphaXCircle.vue';
  import CheckCircleIcon from 'vue-material-design-icons/CheckCircle.vue';
  import { computed, getCurrentInstance, PropType } from "vue";
  import { Answer } from "../types/answer.ts";

  const { emit } = getCurrentInstance()

  const isCorrect = computed(() => {
    if (props.correctAnswerId === 0) {
      return 'answer'
    }
    if (props.alternative.id === props.correctAnswerId) {
      return 'correct'
    } else {
      return 'incorrect'
    }
  })

  const props = defineProps({
    alternative: {
      type: Object as PropType<Answer>,
      required: true
    },

    selected: {
      type: Boolean,
      required: true,
      default: false
    },

    correctAnswerId: {
      type: Number,
      required: true,
      default: 0
    }
  })

  defineEmits(['click'])
</script>

<style scoped lang="scss">
  .question-item {
    border: 1px solid $color-border-answer;
    border-radius: 24px;
    display: flex;
    align-items: center;
    justify-content: space-between;
    box-sizing: border-box;
    padding-left: 30px;
    background-color: $bg-answer;
    cursor: pointer;
    width: 100%;
    height: 55px;

    &.selected {
      border: 2px solid $color-border-selected-answer;
      background-color: $bg-selected-answer;

      &.correct {
        border: 3px solid $color-border-correct-answer;
      }
    }

    &.correct {
      border: 3px solid $color-border-wrong-answer;
    }

    div:nth-child(2) {
      display: flex;
    }
  }
</style>