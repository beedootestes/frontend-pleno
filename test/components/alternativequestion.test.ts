import { describe, expect, test} from "vitest";
import AlternativeQuestion from "../../src/components/AlternativeQuestion.vue";
import {mount, VueWrapper} from "@vue/test-utils";
import CheckCircleIcon from "vue-material-design-icons/CheckCircle.vue";


describe('Test AlternativeQuestion.vue', () => {
  test('mount component', async () => {
    expect(AlternativeQuestion).toBeTruthy()
    const wrapper: VueWrapper<AlternativeQuestion> = mount(AlternativeQuestion, {
      props: {
        alternative: {
          "id": 57492976,
          "text": "Alternative question.",
          "question_id": 72648,
          "is_correct": true
        },
        selected: false,
        correctAnswerId: 0
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('test if emit a click when clicked', async () => {
    const wrapper: VueWrapper<AlternativeQuestion> = mount(AlternativeQuestion, {
      props: {
        alternative: {
          "id": 57492976,
          "text": "Alternative question.",
          "question_id": 72648,
          "is_correct": true
        },
        selected: false,
        correctAnswerId: 0
      }
    })

    await wrapper.trigger('click')
    expect(wrapper.emitted('click'))
  })

  test('test if when selected alter class style', async () => {
    const wrapper: VueWrapper<AlternativeQuestion> = mount(AlternativeQuestion, {
      props: {
        alternative: {
          "id": 57492976,
          "text": "Alternative question.",
          "question_id": 72648,
          "is_correct": true
        },
        selected: false,
        correctAnswerId: 0
      }
    })

    expect(wrapper.attributes('class')).not.contain('selected')
    await wrapper.setProps({ selected: true })
    expect(wrapper.attributes('class')).contain('selected')
  })


  test('test correctAnswerId props alterations', async () => {
    const wrapper: VueWrapper<AlternativeQuestion> = mount(AlternativeQuestion, {
      props: {
        alternative: {
          "id": 57492976,
          "text": "Alternative question.",
          "question_id": 72648,
          "is_correct": true
        },
        selected: false,
        correctAnswerId: 0
      }
    })

    expect(wrapper.attributes('class')).contain('answer')
    await wrapper.setProps({ correctAnswerId: 57492976 })
    expect(wrapper.attributes('class')).contain('correct')
    await wrapper.setProps({ correctAnswerId: 57492987 })
    expect(wrapper.attributes('class')).contain('incorrect')
  })

  test('test show icons conform answer', async () => {
    const wrapper: VueWrapper<AlternativeQuestion> = mount(AlternativeQuestion, {
      props: {
        alternative: {
          "id": 57492976,
          "text": "Alternative question.",
          "question_id": 72648,
          "is_correct": true
        },
        selected: false,
        correctAnswerId: 0
      }
    })

    expect(wrapper.findComponent({ name: "CheckCircleIcon" }).exists()).toBe(false)
    expect(wrapper.findComponent({ name: "AlphaXCircleIcon" }).exists()).toBe(false)

    await wrapper.setProps({ correctAnswerId: 57492976 })
    expect(wrapper.findComponent({ name: "CheckCircleIcon" }).exists()).toBe(true)
    expect(wrapper.findComponent({ name: "AlphaXCircleIcon" }).exists()).toBe(false)

    await wrapper.setProps({ correctAnswerId: 57492987 })
    expect(wrapper.findComponent({ name: "CheckCircleIcon" }).exists()).toBe(false)
    expect(wrapper.findComponent({ name: "AlphaXCircleIcon" }).exists()).toBe(true)
  })
})