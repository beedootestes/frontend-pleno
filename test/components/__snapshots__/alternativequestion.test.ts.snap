// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test AlternativeQuestion.vue > mount component 1`] = `
"<div data-v-58d5336b=\\"\\" class=\\"answer question-item\\">
  <div data-v-58d5336b=\\"\\">Alternative question.</div>
  <div data-v-58d5336b=\\"\\">
    <!--v-if-->
  </div>
</div>"
`;
